File: cpp_ai/tests/test_bpqueue.cpp
#include "doctest.h"
#include "bpqueue.hpp"

using namespace cpp_ai;

TEST_CASE("BPQueueTest - Constructor") {
    BPQueue bpq(-3, 3);
    CHECK(bpq.is_empty());
    CHECK(bpq.get_max() == -4); // a - 1
}

TEST_CASE("BPQueueTest - AppendAndPop") {
    BPQueue bpq(-5, 5);
    Dllink<std::array<int, 2>> a({0, 1});
    Dllink<std::array<int, 2>> b({0, 2});
    Dllink<std::array<int, 2>> c({0, 3});

    bpq.append(&a, 3);
    bpq.append(&b, -2);
    bpq.append(&c, 5);

    CHECK_FALSE(bpq.is_empty());
    CHECK(bpq.get_max() == 5);

    auto* item = bpq.popleft();
    CHECK(item->data()[1] == 3);
    CHECK(bpq.get_max() == 3);

    item = bpq.popleft();
    CHECK(item->data()[1] == 1);
    CHECK(bpq.get_max() == -2);

    item = bpq.popleft();
    CHECK(item->data()[1] == 2);
    CHECK(bpq.is_empty());
}

TEST_CASE("BPQueueTest - Appendlft") {
    BPQueue bpq(-5, 5);
    Dllink<std::array<int, 2>> a({0, 1});
    Dllink<std::array<int, 2>> b({0, 2});

    bpq.appendleft(&a, 3);
    bpq.appendleft(&b, 3);

    auto* item1 = bpq.popleft();
    CHECK(item1->data()[1] == 2); // b was appended left, so it comes first

    auto* item2 = bpq.popleft();
    CHECK(item2->data()[1] == 1); // a comes second
}

TEST_CASE("BPQueueTest - Clear") {
    BPQueue bpq(-5, 5);
    Dllink<std::array<int, 2>> node({0, 1});
    bpq.append(&node, 3);
    bpq.clear();
    CHECK(bpq.is_empty());
}

TEST_CASE("BPQueueTest - KeyManipulation") {
    BPQueue bpq(-5, 5);
    Dllink<std::array<int, 2>> a({0, 1});

    bpq.append(&a, 0);
    CHECK(bpq.get_max() == 0);

    bpq.increase_key(&a, 2);
    CHECK(bpq.get_max() == 2);

    bpq.decrease_key(&a, 3);
    CHECK(bpq.get_max() == -1);

    bpq.modify_key(&a, 4);
    CHECK(bpq.get_max() == 3);

    bpq.modify_key(&a, -5);
    CHECK(bpq.get_max() == -2);
}

TEST_CASE("BPQueueTest - Detach") {
    BPQueue bpq(-5, 5);
    Dllink<std::array<int, 2>> a({0, 1});
    Dllink<std::array<int, 2>> b({0, 2});

    bpq.append(&a, 3);
    bpq.append(&b, 5);

    bpq.detach(&a);
    CHECK(bpq.get_max() == 5);

    auto* item = bpq.popleft();
    CHECK(item->data()[1] == 2);
    CHECK(bpq.is_empty());
}

TEST_CASE("BPQueueTest - LockedItem") {
    BPQueue bpq(-5, 5);
    Dllink<std::array<int, 2>> a({0, 1});
    bpq.append(&a, 0);
    a.lock();
    bpq.modify_key(&a, 3); // Should have no effect
    CHECK(bpq.get_max() == 0);
}

File: rust_ai/src/tests/test_bpqueue.rs
use crate::bpqueue::BPQueue;
use crate::dllist::Dllink;

#[test]
fn test_bpqueue_constructor() {
    let bpq = BPQueue::new(-3, 3);
    assert!(bpq.is_empty());
    assert_eq!(bpq.get_max(), -4); // a - 1
}

#[test]
fn test_bpqueue_append_and_pop() {
    let mut bpq = BPQueue::new(-5, 5);
    let mut a = Dllink::new([0, 1]);
    let mut b = Dllink::new([0, 2]);
    let mut c = Dllink::new([0, 3]);
    a.lock();
    b.lock();
    c.lock();

    unsafe {
        bpq.append(&mut a, 3);
        bpq.append(&mut b, -2);
        bpq.append(&mut c, 5);
    }

    assert!(!bpq.is_empty());
    assert_eq!(bpq.get_max(), 5);

    unsafe {
        bpq.popleft();
        assert_eq!(bpq.get_max(), 3);

        bpq.popleft();
        assert_eq!(bpq.get_max(), -2);

        bpq.popleft();
    }
    assert!(bpq.is_empty());
}

#[test]
fn test_bpqueue_appendleft() {
    let mut bpq = BPQueue::new(-5, 5);
    let mut a = Dllink::new([0, 1]);
    let mut b = Dllink::new([0, 2]);

    unsafe {
        bpq.appendleft(&mut a, 3);
        bpq.appendleft(&mut b, 3);
    }

    unsafe {
        bpq.popleft();
        bpq.popleft();
    }
}

#[test]
fn test_bpqueue_clear() {
    let mut bpq = BPQueue::new(-5, 5);
    let mut a = Dllink::new([0, 1]);

    unsafe {
        bpq.append(&mut a, 3);
    }
    bpq.clear();
    assert!(bpq.is_empty());
}

#[test]
fn test_bpqueue_key_manipulation() {
    let mut bpq = BPQueue::new(-5, 5);
    let mut a = Dllink::new([0, 1]);

    unsafe {
        bpq.append(&mut a, 0);
        assert_eq!(bpq.get_max(), 0);

        bpq.increase_key(&mut a, 2);
        assert_eq!(bpq.get_max(), 2);

        bpq.decrease_key(&mut a, 3);
        assert_eq!(bpq.get_max(), -1);

        bpq.modify_key(&mut a, 4);
        assert_eq!(bpq.get_max(), 3);

        bpq.modify_key(&mut a, -5);
        assert_eq!(bpq.get_max(), -2);
    }
}

#[test]
fn test_bpqueue_detach() {
    let mut bpq = BPQueue::new(-5, 5);
    let mut a = Dllink::new([0, 1]);
    let mut b = Dllink::new([0, 2]);

    unsafe {
        bpq.append(&mut a, 3);
        bpq.append(&mut b, 5);

        bpq.detach(&mut a);
        assert_eq!(bpq.get_max(), 5);

        bpq.popleft();
    }
    assert!(bpq.is_empty());
}

#[test]
fn test_bpqueue_locked_item() {
    let mut bpq = BPQueue::new(-5, 5);
    let mut a = Dllink::new([0, 1]);

    unsafe {
        bpq.append(&mut a, 0);
        a.lock();
        bpq.modify_key(&mut a, 3); // Should have no effect
        assert_eq!(bpq.get_max(), 0);
    }
}
